### PR TITLE
fix using *_Prev_iter(it) instead of *(it - 1)

### LINKED ISSUE #465: <algorithm>: Consider using *_Prev_iter(it) instead of *(it - 1)
We have a helper function that's a simplified form of `std::prev`:

https://github.com/microsoft/STL/blob/fd04f77dd217a7502c77e0cbf45e24fcb84ed17c/stl/inc/xutility#L1638-L1641

We also have algorithms that need to access the element immediately before where an iterator points to:

https://github.com/microsoft/STL/blob/fd04f77dd217a7502c77e0cbf45e24fcb84ed17c/stl/inc/algorithm#L3459

It may be more efficient to use `_Prev_iter` for user-defined iterators where decrementing is cheaper than general-purpose subtraction. There are several occurrences in the STL (more after #464 is merged). Currently, I see no occurrences of `+ 1` that could benefit from `_Next_iter`.

### PR DESCRIPTION
**Pull request by JeanPhilippeKernel**
# Description

Fixes #465 

# Checklist

Be sure you've read README.md and understand the scope of this repo.

If you're unsure about a box, leave it unchecked. A maintainer will help you.

- [x] Identifiers in product code changes are properly `_Ugly` as per
  https://eel.is/c++draft/lex.name#3.1 or there are no product code changes.
- [x] The STL builds successfully and all tests have passed (must be manually
  verified by an STL maintainer before automated testing is enabled on GitHub,
  leave this unchecked for initial submission).
- [ ] These changes introduce no known ABI breaks (adding members, renaming
  members, adding virtual functions, changing whether a type is an aggregate
  or trivially copyable, etc.).
- [x] These changes were written from scratch using only this repository,
  the C++ Working Draft (including any cited standards), other WG21 papers
  (excluding reference implementations outside of proposed standard wording),
  and LWG issues as reference material. If they were derived from a project
  that's already listed in NOTICE.txt, that's fine, but please mention it.
  If they were derived from any other project (including Boost and libc++,
  which are not yet listed in NOTICE.txt), you *must* mention it here,
  so we can determine whether the license is compatible and what else needs
  to be done.

### CODE REVIEW

File: stl/inc/algorithm [hunk 1]
```diff
@@ -4745,7 +4745,8 @@ _CONSTEXPR20 bool next_permutation(_BidIt _First, _BidIt _Last, _Pr _Pred) {
 
     for (;;) { // find rightmost element smaller than successor
         auto _UNext1 = _UNext;
-        if (_DEBUG_LT_PRED(_Pred, *--_UNext, *_UNext1)) { // swap with rightmost element that's smaller, flip suffix
+        if (_DEBUG_LT_PRED(
+                _Pred, *_Prev_iter(_UNext), *_UNext1)) { // swap with rightmost element that's smaller, flip suffix
```
Comment (BillyONeal) [inline stl/inc/algorithm:4749]: This would format better if you insert a newline after the { to put the comment on the following line. (Ditto below)

Comment (JeanPhilippeKernel) [reply stl/inc/algorithm:4749]: Oh okay let me do this then

**Comment by cbezault** [comment]
Yup, it looks like there are four tests in `libc++` which are hanging forever.

**Comment by JeanPhilippeKernel** [comment]
Oh my bad !😯😯, you right ! After that clarification now my changes are totally incorrect!! Thanks a lot !!, 
Let's me UPDATE

File: stl/inc/algorithm [hunk 2]
```diff
@@ -1525,7 +1525,7 @@ _NODISCARD _CONSTEXPR20 _FwdIt search_n(
                 _Iter_diff_t<_FwdIt> _Count1 = _Count_diff;
                 auto _UMid                   = _UFirst;
 
-                while (_UOld_first != _UFirst && _Pred(*(_UFirst - 1), _Val)) { // back up over any skipped prefix
+                while (_UOld_first != _UFirst && _Pred(*_Prev_iter(_UFirst), _Val)) { // back up over any skipped prefix
```
Comment (mnatsuhara) [inline stl/inc/algorithm:1528]: Is there some reason not to also make this change in the `!_HAS_IF_CONSTEXPR` version of this function? (at line 1609)

Comment (JeanPhilippeKernel) [reply stl/inc/algorithm:1528]: thanks @mnatsuhara and @CaseyCarter , forgot this one

**Comment by StephanTLavavej** [comment]
Thanks for the efficiency improvement! Every bit helps. 😺

**Comment by JeanPhilippeKernel** [comment]
😁 it's a pleasure !! thanks y'all !

**Pull request merged**